File: homework4/Satyam_Mishra/Q6_WordBreak.java
// Technique - TRIE + MEMOIZATION
// TIME COMPLEXITY - O(N^2 + WL)
// SPACE COMPLEXITY - O(N^2 + WL)
// where N is length of string, W is number of words in dictionary, L is length of word

package homework4.Satyam_Mishra;

import java.util.HashMap;
import java.util.Map;

public class Q6_WordBreak {
    static class Trie {
        Trie[] children;
        boolean validWord;

        public Trie() {
            children = new Trie[26];
            validWord = false;
        }
    }

    public static void insert(String word, Trie curr) {
        int n = word.length();

        for(int i = 0; i < n; i++) {
            int ch = word.charAt(i) - 'a';

            if(curr.children[ch] == null) {
                curr.children[ch] = new Trie();
            }
            curr = curr.children[ch];
        }

        curr.validWord = true;
    }

    public static boolean canBreak(String word, Trie root, Map<String, Boolean> visited) {
        Trie currNode = root;

        if(word.length() == 0) {
            visited.put(word, true);
            return true;
        }

        if (visited.containsKey(word)) {
            return visited.get(word);
        }

        int n = word.length();

        for(int i = 0; i < n; i++) {
            int ch = word.charAt(i) - 'a';

            if(currNode.children[ch] == null) {
                break;
            }

            currNode = currNode.children[ch];

            if(currNode.validWord) {
                if(canBreak(word.substring(i + 1), root, visited)) {
                    visited.put(word, true);
                    return true;
                }
            }
        }

        visited.put(word, false);
        return false;
    }

    public static void main(String[] args) {
        Trie root = new Trie();

        String[] dictionary = {"elf", "go", "golf", "man", "manatee", "not", "note", "pig", "quip", "tee", "teen"};

        for(String word : dictionary) {
            insert(word, root);
        }
        
        Map<String, Boolean> visited = new HashMap<>();
        System.out.println(canBreak("mangolf", root, visited));
        System.out.println(canBreak("manateenotelf", root, visited));
        System.out.println(canBreak("quipig", root, visited));
    }
}

// time taken - 35m

File: homework4/Aditya_Prasad/q6_WordBreak.cpp
// Technique : DP (Memoization) + Trie as a parameter
// T.C : O(T + n*L)
// M.C : O(T + n) 
// where T is sum of all word lengths, L = max(word[i].size()), n = dict.size()

#include <bits/stdc++.h>
using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    bool isEnd;

    TrieNode() {
        isEnd = false;
        for (int i = 0; i < 26; i++) {
            children[i] = nullptr;
        }
    }
};

class Trie {
public:
    TrieNode* root;

    Trie () {
        root = new TrieNode();
    }

    void insert(string word) {
        TrieNode* node = root;

        for (int i = 0; i < word.size(); i++) {
            int index = word[i] - 'a';
            if (node->children[index] == nullptr) {
                node->children[index] = new TrieNode();
            }

            node = node->children[index];
        }

        node->isEnd = true;
    }

    bool searchWord(string word) {
        TrieNode* node = root;

        for (int i = 0; i < word.size(); i++) {
            int index = word[i] - 'a';
            if (node->children[index] == nullptr) return false;
            node = node->children[index];
        }

        return node->isEnd;
    }
};
bool canBreak(int index, string& word, vector<string>& dictWords, vector<int>& dp, Trie& trie) {
    if (index == word.size()) return true;

    if (dp[index] != -1) return dp[index];

    TrieNode* node = trie.root;

    for (int i = index; i < word.size(); i++) {
        int pos = word[i] - 'a';

        if (node->children[pos] == nullptr) break;

        node = node->children[pos];

        if (node->isEnd == true) {

            bool restResult = canBreak(i + 1, word, dictWords, dp, trie);

            if (restResult == true) {
                return dp[index] = true;
            }
        }
    }

    return dp[index] = false;
}

bool wordBreak(string word, vector<string>& dictWords) {
    Trie trie;
    for (int i = 0; i < dictWords.size(); i++) {
        trie.insert(dictWords[i]);
    }
    vector<int> dp(word.size() + 1, -1);
    return canBreak(0, word, dictWords, dp, trie);
}

int main() {
    vector<string> dictWords = {"elf", "go", "golf", "man", "manatee", "not", "note", "pig", "quip", "tee", "teen"};

    cout << wordBreak("golfman", dictWords) << endl;
    cout << wordBreak("manateenotelf", dictWords) << endl;
    cout << wordBreak("quipig", dictWords) << endl;

    return 0;

}

// Time Taken : 38 mins